nixos/lib/test-driver/test_driver/qmp.py

```diff
@@ -0,0 +1,65 @@
+import socket
+from queue import Queue
+from typing import Dict, Any, Iterator
+import json
+from pathlib import Path
+
+
+class QMPSession:
+    def __init__(self, sock: socket.socket) -> None:
+        self.sock = sock
+        self.pending_events: Queue[Dict[str, Any]] = Queue()
+        self.reader = sock.makefile("r")
+        self.writer = sock.makefile("w")
+        hello = self._result()
+        assert "QMP" in hello, f"Unexpected result: {hello}"
+        self.send("qmp_capabilities")
+
+    @classmethod
+    def from_path(cls, path: Path) -> "QMPSession":
+        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+        sock.connect(str(path))
+        return cls(sock)
+
+    def __del__(self) -> None:
+        self.sock.close()
+
+    def _readmsg(self) -> Dict[str, Any]:
+        line = self.reader.readline()
+        return json.loads(line)
+
+    def _raise_unexpected_msg(self, msg: Dict[str, Any]) -> None:
+        m = json.dumps(msg, sort_keys=True, indent=4)
+        raise RuntimeError(f"Got unexpected qmp response: {m}")
+
+    def _result(self) -> Dict[str, Any]:
+        while True:
+            # QMP is in the handshake
+            res = self._readmsg()
+            if "return" in res or "QMP" in res:
+                return res
+            elif "event" in res:
+                self.pending_events.put(res)
+                continue
+            else:
+                self._raise_unexpected_msg(res)
+
+    def events(self) -> Iterator[Dict[str, Any]]:
+        while not self.pending_events.empty():
+            yield self.pending_events.get()
+
+        res = self._readmsg()
+
+        if "event" not in res:
+            self._raise_unexpected_msg(res)
+        yield res
+
+    def send(self, cmd: str, args: Dict[str, str] = {}) -> Dict[str, str]:
+        data: Dict[str, Any] = dict(execute=cmd)
+        if args != {}:
+            data["arguments"] = args
+
+        json.dump(data, self.writer)
+        self.writer.write("\n")
+        self.writer.flush()
+        return self._result()
```

Comment:
Currently, it seems like users would have to care themselves that there are no pending results... is that correct?

this code would look much safer to me if we simply had a results queue and an event queue that both are refilled from a shake_the_tree function. this code could then:

- shake the tree
- check that there is currently no unqueried result (or simply drop them)
- send the message
- shake the tree again (maybe in a loop until there is a result)
- return the result

Reply:
I went for an assert rather than dropping them, let me know if I implemented enough of your proposal, the shaking is called `read_pending_messages` and is public, the "wait for new result" is `_wait_for_new_result`, I feel it's kinda an impl detail that should not be offered to public consumers.